calib/accorth.c

```diff
@@ -0,0 +1,503 @@
+/*
+* Data acquisition for earth acceleration towards local frame of reference versors.
+* Accelerations are saved to `tab[6]` in the following axis directions: { x+, x-, y+, y-, z+, z- }.
+*/
+static void accorth_sixPtsCalib(vec_t tab[6])
+{
+	printf("Place the drone as specified. Positions should be very stable (rather than precise).\n");
+	fflush(stdin);
+
+	printf("Place drone: %s\n", "flat on the ground");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[5], 1000); /* flat on the ground should give earth acceleration towards negative 'z' */
+
+	printf("Place drone: %s\n", "flat on the ground, belly up\n");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[4], 1000); /* flat, belly up should give earth acceleration towards positive 'z' */
+
+	printf("Place drone: %s\n", "on the left side\n");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[2], 1000); /* left down should give earth acceleration towards positive 'y' */
+
+	printf("Place drone: %s\n", "on the right side\n");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[3], 1000); /* left down should give earth acceleration towards negative 'y' */
+
+	printf("Place drone: %s\n", "nose down\n");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[1], 1000); /* nose down should give earth acceleration towards negative 'x' */
+
+	printf("Place drone: %s\n", "nose up\n");
+	printf("Press [Enter] to sample");
+	fflush(stdout);
+	getchar();
+	accorth_accelAvg(&tab[0], 1000); /* nose down should give earth acceleration towards positive 'x' */
+}
+
+
+/* Calculating S3 and H3 from S1, S2, H1, H2. Returns -1 on calculation error, 0 otherwise. */
+static int accorth_paramsCombine(const matrix_t *S1, const matrix_t *H1, const matrix_t *S2, const matrix_t *H2, matrix_t *S3, matrix_t *H3)
+{
+	float invBuf[18], dataTmp[3];
+	matrix_t tmp = { .data = dataTmp, .rows = 3, .cols = 1, .transposed = 0 };
+
+	/* Using S3 as temporary storage for inv(S1) */
+	if (matrix_inv(S1, S3, invBuf, 18) < 0) {
+		printf("Cannot calculate S1 inverse\n");
+		return -1;
+	}
+
+	matrix_prod(S3, H2, &tmp);
+	matrix_add(&tmp, H1, H3);
+	matrix_prod(S2, S1, S3);
+
+	return 0;
+}
+
+
+/* Fits ellipsoid to `n` samples in `data` and stores calibration parameters to `S` and `H` */
+static int accorth_ellipsoidFit(vec_t *data, size_t n, matrix_t *S, matrix_t *H)
+{
+	fit_lma_t lma;
+	int i;
+
+	if (lma_init(3, 12, n, ellcal_lmaJacobian, ellcal_lmaResiduum, ellcal_lmaGuess, &lma) < 0) {
+		fprintf(stderr, "%s: failed to init LMA", ACCORTH_TAG);
+		return -1;
+	}
+
+	/* Write measurements to lma `samples` matrix */
+	for (i = 0; i < n; i++) {
+		MATRIX_DATA(&lma.samples, i, 0) = accorth_common.meas[i].x;
+		MATRIX_DATA(&lma.samples, i, 1) = accorth_common.meas[i].y;
+		MATRIX_DATA(&lma.samples, i, 2) = accorth_common.meas[i].z;
```

Comment:
It should be:
```suggestion
		MATRIX_DATA(&lma.samples, i, 0) = data[i].x;
		MATRIX_DATA(&lma.samples, i, 1) = data[i].y;
		MATRIX_DATA(&lma.samples, i, 2) = data[i].z;
```
corrected in an additional commit on this branch